build_modules/lib/src/meta_module.dart

```diff
@@ -0,0 +1,433 @@
+import 'dart:async';
+import 'dart:collection';
+import 'dart:math';
+
+import 'package:analyzer/analyzer.dart';
+import 'package:build/build.dart';
+import 'package:path/path.dart' as p;
+import 'package:json_annotation/json_annotation.dart';
+
+import 'modules.dart';
+
+part 'meta_module.g.dart';
+
+/// Returns the top level directory in [uri].
+///
+/// Throws an [ArgumentError] if [uri] is just a filename with no directory.
+String _topLevelDir(String uri) {
```

Comment:
nit: rename argument to `path` since its not a `Uri`. I think we already have this implemented somewhere as well.

Reply:
Done. Couldn't find any other implementation after some basic searches.